lib/mayaUsd/ufe/MayaUsdContextOps.cpp

```diff
@@ -650,53 +714,106 @@ Ufe::UndoableCommand::Ptr MayaUsdContextOps::doOpCmd(const ItemPath& itemPath)
         }
         return compositeCmd;
     } else if (itemPath[0] == UnbindMaterialUndoableCommand::commandName) {
-        return std::make_shared<UnbindMaterialUndoableCommand>(fItem->path());
+        return std::make_shared<UnbindMaterialUndoableCommand>(_item->path());
 #ifdef UFE_V4_FEATURES_AVAILABLE
     } else if (itemPath.size() == 3u && itemPath[0] == kAssignNewMaterialItem) {
-        // Make a copy so that we don't change the user's original selection.
-        Ufe::Selection sceneItems(*Ufe::GlobalSelection::get());
-        // As per UX' wishes, we add the item that was right-clicked,
-        // regardless of its selection state.
-        sceneItems.append(fItem);
-        if (sceneItems.size() > 0u) {
-            return std::make_shared<InsertChildAndSelectCommand>(
-                UsdUndoAssignNewMaterialCommand::create(sceneItems, itemPath[2]));
-        }
+        // In single context item mode, only assign material to the context item.
+        return std::make_shared<InsertChildAndSelectCommand>(
+            UsdUndoAssignNewMaterialCommand::create(_item, itemPath[2]));
     } else if (itemPath.size() == 3u && itemPath[0] == kAddNewMaterialItem) {
         return std::make_shared<InsertChildAndSelectCommand>(
-            UsdUndoAddNewMaterialCommand::create(fItem, itemPath[2]));
+            UsdUndoAddNewMaterialCommand::create(_item, itemPath[2]));
     } else if (itemPath.size() == 3u && itemPath[0] == kAssignExistingMaterialItem) {
-        std::shared_ptr<Ufe::CompositeUndoableCommand> compositeCmd;
-        Ufe::Selection                                 sceneItems(*Ufe::GlobalSelection::get());
-        sceneItems.append(fItem);
-        for (auto& sceneItem : sceneItems) {
-            if (BindMaterialUndoableCommand::CompatiblePrim(sceneItem)) {
-                if (!compositeCmd) {
-                    compositeCmd = std::make_shared<Ufe::CompositeUndoableCommand>();
+        // In single context item mode, only assign material to the context item.
+        return std::make_shared<BindMaterialUndoableCommand>(_item->path(), SdfPath(itemPath[2]));
+#endif
+    }
+    return nullptr;
+}
+
+Ufe::UndoableCommand::Ptr MayaUsdContextOps::doBulkOpCmd(const ItemPath& itemPath)
+{
+    // First check if our base class handles this item.
+    auto cmd = Parent::doBulkOpCmd(itemPath);
+    if (cmd)
+        return cmd;
+
+    // Create the composite undoable command here and fill it below.
+    // If the composite command is empty (meaning we didn't add any
+    // commands above) return nullptr instead (so nothing will be executed).
+    auto compositeCmd = std::make_shared<Ufe::CompositeUndoableCommand>();
+
+#ifdef DEBUG
+    auto DEBUG_OUTPUT = [&compositeCmd](const Ufe::Selection& bulkItems) {
+        TF_STATUS(
+            "Performing bulk edit on %d prims (%d selected)",
+            compositeCmd->cmdsList().size(),
+            bulkItems.size());
+    };
+#else
+#define DEBUG_OUTPUT(x) (void)0
+#endif
```

Comment:
Helpful debugging output to show how many bulk items there were and how many the menu action was applied to. Only outputs in debug builds.